### PR TITLE
Image Segmentation pipeline

### PR DESCRIPTION
**Pull request by mishig25**
# What does this PR do?
## TLDR
Implements `image-segmentation` pipeline (for [DetrForSegmentation](https://huggingface.co/transformers/model_doc/detr.html#detrforsegmentation) atm). 
## API specifications
Input: image source (identical to the input of `image-classification` & `object-detection` pipelines)
Output:
```python
List(Dict(
    mask: str, # base64 str
    label: float,
    score: int,
))
```

### Two design choices I've made and would like to discuss (& modify if needed):
1. Output `png_string` has masks information using same mechanism as COCO panoptic segmentation annotations. See section `4. Panoptic Segmentation` from https://cocodataset.org/#format-data. Paraphrasing a bit:
> per-pixel segment ids are stored in the PNG string. Each segment is assigned a unique id. Unlabeled pixels (void) are assigned a value of 0. Note that when you load the PNG as an RGB image, you will need to compute the ids via ids=R+G*256+B*256^2.
2. Image segmentation pipeline accepts `subtask` arg. There are different variations of segmentation task (semantic, instance, panoptic, etc. see image below). If a model doesn't implement requested subtask, it gets defaulted to what's available. See example below:
https://github.com/huggingface/transformers/blob/dd5c2697f129988bd9c70b52555e49dd32c78bd2/src/transformers/models/detr/feature_extraction_detr.py#L738-L739

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      Discussed in huggingface/hub-docs#43, huggingface/hub-docs#6, https://github.com/huggingface/huggingface_hub/pull/378
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), 
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

#### Update 1:
Updated output shape:
```python
List(Dict(
    mask: str, # base64 str
    label: float,
    score: int,
))
```
The goal for any segmentation architecture is to implement the most detailed version of segmentation subtask they can so that other subtasks can be reconstructed (if needed). For example, if an architecture post_process_segmentation method implements part-aware panoptic, other subtasks (including semantic, instance, etc.) can be reconstructed from part-aware panoptic output since all the details needed are in there

### CODE REVIEW

File: tests/test_pipelines_image_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+                        "label": ANY(str),
+                    },
+                )
+
+        image_segmenter = ImageSegmentationPipeline(model=model, feature_extractor=feature_extractor)
+        outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0)
+        test_outputs(outputs)
+
+        subtasks = [None, "panoptic", "instance", "semantic"]
+        for subtask in subtasks:
```
Comment (Narsil) [inline tests/test_pipelines_image_segmentation.py:78]: Never use for loops in tests. Or very exceptionally.
Never use inner functions in tests.

Both really hurt the main asset of a test, which is readability. A reader needs to be able to read a single test and know if it's valid or not, with as little as possible surrounding context.

Here, reader needs to understand what variable is sent to `test_outputs`, understand the function itself, then reason if variable was supposed to be A, [A] or something else to assess if the test is correct.


```python
outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0)
self.assertEqual(nested_simply(outputs, {.....})

# Easy to read that if input is list, output is list.
outputs = image_segmenter(["./tests/fixtures/tests_samples/COCO/000000039769.png"], threshold=0.0)
self.assertEqual(nested_simply(outputs, [{.....}])

outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0, subtask="panoptic")
self.assertEqual(nested_simply(outputs, {.....})

....
```

File: tests/test_pipelines_image_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        model_id = "mishig/tiny-detr-mobilenetsv3-panoptic"
+
+        model = AutoModelForImageSegmentation.from_pretrained(model_id)
+        feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/detr-resnet-50-panoptic")
+        image_segmenter = ImageSegmentationPipeline(model=model, feature_extractor=feature_extractor)
+
+        outputs = image_segmenter("http://images.cocodataset.org/val2017/000000039769.jpg", threshold=0.0)
+
+        # shortening by hashing
+        outputs["png_string"] = hashlib.sha1(outputs["png_string"].encode("UTF-8")).hexdigest()
```
Comment (Narsil) [inline tests/test_pipelines_image_segmentation.py:120]: Neat trick to keep test both readable and enforce correctness !

File: tests/test_pipelines_image_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+        outputs = image_segmenter(
+            [
+                "http://images.cocodataset.org/val2017/000000039769.jpg",
+                "http://images.cocodataset.org/val2017/000000039769.jpg",
+            ],
+            threshold=0.0,
+        )
+        for o in outputs:
+            o["png_string"] = hashlib.sha1(o["png_string"].encode("UTF-8")).hexdigest()
```
Comment (Narsil) [inline tests/test_pipelines_image_segmentation.py:139]: Here is an acceptable exception to the no forloop in tests point I was making.

File: src/transformers/pipelines/image_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    def _forward(self, model_inputs):
+        target_size = model_inputs.pop("target_size")
+        outputs = self.model(**model_inputs)
+        model_outputs = {"outputs": outputs, "target_size": target_size}
+        return model_outputs
+
+    def postprocess(self, model_outputs, threshold=0.9, subtask=None):
+        raw_annotations = self.feature_extractor.post_process_segmentation(
+            model_outputs["outputs"], model_outputs["target_size"], threshold=threshold, subtask=subtask
```
Comment (Narsil) [inline src/transformers/pipelines/image_segmentation.py:135]: `subtask` will be generic here and work for ALL models, meaning ALL `feature_extractor` for images that support segmentation needs to accept the `subtask` kwarg (and all its variants).

I am not sure we desire this. Can't we just use `part-aware` all the time for `detr` as it's the most general (If I understand correctly) ?

Other option would be to ask the model it's fullest possible description and do something within the pipeline to actually recover a common representation (`token-classification` 's `aggregate_strategy` is an example of what I mean)

Looking at the proposed changes for masks, I am pretty sure we can drop `subtask` altogether.

File: tests/test_pipelines_image_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        )
+
+    @require_torch
+    @slow
+    def test_subtask(self):
+        model_id = "facebook/detr-resnet-50-panoptic"
+
+        image_segmenter = pipeline("image-segmentation", model=model_id)
+
+        outputs = image_segmenter("http://images.cocodataset.org/val2017/000000039769.jpg", subtask="panoptic")
```
Comment (Narsil) [inline tests/test_pipelines_image_segmentation.py:228]: This is EXACTLY the same as no `subtask`. What is the point of having subtasks ?

**Review by Narsil — COMMENTED** [review]
Very nice PR! 

I think we can add more representative power to the pipeline output removing the need for `subtask` or at least internalize it's complexity (so we don't force model implementors to add irrelevant stuff)

The tests can be simplified quite a bit IMO.
If we really want `subtask` to exist, we need test that display the differences they imply.

Happy to help on changing the representation if needed.

File: src/transformers/pipelines/image_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+              segment ids are stored as the PNG image. Each segment is assigned a unique id. Unlabeled pixels (void)
+              are assigned a value of 0. Note that when you load the PNG as an RGB image, you will need to compute the
+              ids via ids=R+G*256+B*256^2 (inspired by https://cocodataset.org/#format-data).
+            - **segments_info** (:obj:`List[dict]`) -- list that contains labels for segments.
+
+            **segments_info** contains dictionaries with the following keys:
+
+            - **id** (:obj:`int`) -- The id attributed for segment.
+            - **label** (:obj:`str`) -- The class label identified by the model.
+            - **score** (:obj:`float`) -- The score attributed by the model for that label.
```
[the remark this answers is not in the document]
Comment (mishig25) [reply src/transformers/pipelines/image_segmentation.py:115]: Am I correct to assume that you are proposing the output to be: 
```py
List(Dict(
    mask: str, // base64 str
    label: float,
    score: number,
    id: int
))
```
?

[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/image_segmentation.py:115]: Exactly ! I might be missing some subtleties but it should work.

(For `part-aware` for instance you need some kind of tree structure, but I think `parent?: int` could work in that case where `parent` will be the index of the parent mask).

File: src/transformers/models/auto/modeling_auto.py [hunk 7]
```diff
@@ -228,6 +228,13 @@
     ]
 )
 
+MODEL_FOR_IMAGE_SEGMENTATION_MAPPING_NAMES = OrderedDict(
+    [
+        # Model for Image Segmentation mapping
+        ("detr", "DetrForSegmentation"),
```
Comment (NielsRogge) [inline src/transformers/models/auto/modeling_auto.py:234]: `DetrForSegmentation` was probably not the best name. Could we add an alias called `DetrForImageSegmentation`? Or should we distinguish between the different kinds of segmentation (this particular one is panoptic segmentation)

File: src/transformers/pipelines/image_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        postprocess_kwargs = {}
+        if "threshold" in kwargs:
+            postprocess_kwargs["threshold"] = kwargs["threshold"]
+        if "subtask" in kwargs:
+            postprocess_kwargs["subtask"] = kwargs["subtask"]
+        return {}, {}, postprocess_kwargs
+
+    def __call__(self, *args, **kwargs) -> Union[Predictions, List[Prediction]]:
+        """
+        Perform segmentation (detect maskss & classes) in the image(s) passed as inputs.
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:81]:
```suggestion
        Perform segmentation (detect masks & classes) in the image(s) passed as inputs.
```

File: src/transformers/pipelines/image_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+
+Prediction = Dict[str, Any]
+Predictions = List[Prediction]
+
+
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class ImageSegmentationPipeline(Pipeline):
+    """
+    Image segmntation pipeline using any :obj:`AutoModelForImageSegmentation`. This pipeline predicts masks of objects
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:30]:
```suggestion
    Image segmentation pipeline using any :obj:`AutoModelForImageSegmentation`. This pipeline predicts masks of objects
```

File: src/transformers/pipelines/image_segmentation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mishig25) [reply src/transformers/pipelines/image_segmentation.py:115]: If I'm not mistaken (cc: @NielsRogge), the current DetrForSegmentation implements `panoptic` segmentation, not `part-aware panoptic`. Therefore, should I just put -1 the parent indeces for DetrForSegmentation atm?
```py
List(Dict(
    mask: str, // base64 str
    label: float,
    score: int,
    parent: int // -1 for all for DetrForSegmentation
))
# removing `id` field from dict since it is no longer needed
```
And in the future, we might have segmentation architectures that can do `part-aware panoptic`, in which case `parent` fields will come in handy. WDYT @Narsil

File: src/transformers/models/auto/modeling_auto.py (hunk 7 above)
Comment (mishig25) [reply src/transformers/models/auto/modeling_auto.py:234]: If I'm not mistaken, as noted in [this discussion](https://github.com/huggingface/transformers/pull/13828#discussion_r721652671), we will not make any distinction between different segmentation subtasks and expect models/architectures to implement post_process_segmentation method that implements "the most detailed version of subtask" they can.

More details:
This output shape
```py
List(Dict(
    mask: str, // base64 str
    label: float,
    score: int,
    parent: int // -1 for all for DetrForSegmentation
))
# removing `id` field from dict since it is no longer needed
```
will allow us to implement `part-aware panoptic` segmentation (which is the most detailed segmentation subtask). And if using this output, a user can construct `semantic`, `instance` or any othre segmentation subtask if they want

File: src/transformers/pipelines/image_segmentation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/image_segmentation.py:115]: Yes, let's not add what's not necessary, we know how to add that's enough (a commentary maybe for future implementors but that's the extent of where I would go.

[the remark this answers is not in the document]
Comment (mishig25) [reply src/transformers/pipelines/image_segmentation.py]: pushed the suggested changes:
1. Change the output shape to the one discussed above
2. Remove subtask argument, thereby making tests simpler

**Review by Narsil — APPROVED** [review]
LGTM !

**Comment by Narsil** [comment]
It's gtg for me.
